Sources/NIOSOCKS/Messages/ClientRequest.swift

```diff
@@ -0,0 +1,188 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+// MARK: - ClientRequest
+
+/// Instructs the SOCKS proxy server of the target host,
+/// and how to connect.
+struct ClientRequest: Hashable {
+    
+    /// The SOCKS protocol version - we currently only support v5.
+    public let version: UInt8 = 5
+    
+    /// How to connect to the host.
+    public var command: Command
+    
+    /// The target host address.
+    public var addressType: AddressType
+    
+    /// Creates a new `ClientRequest`.
+    /// - parameter command: How to connect to the host.
+    /// - parameter addressType: The target host address.
+    /// - parameter desiredPort: The target host port.
+    public init(command: Command, addressType: AddressType) {
+        self.command = command
+        self.addressType = addressType
+    }
+    
+}
+
+extension ByteBuffer {
+    
+    @discardableResult mutating func writeClientRequest(_ request: ClientRequest) -> Int {
+        var written = 0
+        written += self.writeInteger(request.version)
+        written += self.writeInteger(request.command.value)
+        written += self.writeInteger(UInt8(0))
+        written += self.writeAddressType(request.addressType)
+        return written
+    }
+    
+}
+
+// MARK: - Command
+
+/// What type of connection the SOCKS server should establish with
+/// the target host.
+public struct Command: Hashable {
+    
+    /// Typically the primary connection type, suitable for HTTP.
+    public static let connect = Command(value: 0x01)
+    
+    /// Used in protocols that require the client to accept connections
+    /// from the server, e.g. FTP.
+    public static let bind = Command(value: 0x02)
+    
+    /// Used to establish an association within the UDP relay process to
+    /// handle UDP datagrams.
+    public static let udpAssociate = Command(value: 0x03)
+    
+    public var value: UInt8
+}
+
+// MARK: - AddressType
+
+/// The address used to connect to the target host.
+public enum AddressType: Hashable {
+    
+    case address(SocketAddress)
+    
+    case domain(String, port: UInt16)
```

Comment:
For consistency with NIO, this should be:

```suggestion
    case domain(String, port: Int)
```